### PR TITLE
Provide completion for cache name in configuration file

### LINKED ISSUE #404: Provide completion for cache name in configuration file
If your code contains 
```java
@CacheResult(cacheName = "foo")
```

Then completion for cache configuration should be able to automatically propose the proper names in application.properties:
```
quarkus.cache.caffeine.foo.expire-after-access
quarkus.cache.caffeine.foo.expire-after-write
quarkus.cache.caffeine.foo.maximum-size
quarkus.cache.caffeine.foo.initial-capacity
```
 instead of `{*}`:

Screenshot 2021-07-21 at 10 33 56

See https://quarkus.io/guides/cache#caffeine-configuration-properties

### PR DESCRIPTION
**Pull request by angelozerr**
Provide completion for cache name in configuration file

Fixes #404

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
It should starts working. You can play with https://github.com/quarkusio/quarkus-quickstarts/tree/main/cache-quickstart

I need to clean my code and write tests.

**Comment by angelozerr** [comment]
This PR requires the LSP4MP PR https://github.com/eclipse/lsp4mp/pull/159 which fixes the bug of disapear of quarkus cache properties as soon as you update a Java file.

**Comment by angelozerr** [comment]
Tests has been writen.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/docker/Dockerfile.legacy-jar [hunk 1]
```diff
@@ -0,0 +1,51 @@
+####
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/docker/Dockerfile.legacy-jar]: fixed

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/docker/Dockerfile.native [hunk 2]
```diff
@@ -0,0 +1,27 @@
+####
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/docker/Dockerfile.native]: fixed

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/resources/META-INF/resources/index.html [hunk 3]
```diff
@@ -0,0 +1,152 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus.test/projects/maven/cache-quickstart/src/main/resources/META-INF/resources/index.html]: fixed

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/cache/properties/QuarkusCacheResultProvider.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/cache/properties/QuarkusCacheResultProvider.java:1]: fixed

File: quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusSearchContext.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	 * 
+	 * @param extensionName the Quarkus extension name.
+	 * @param replacer      the replacer.
+	 */
+	public void registerPropertyMapKeyReplacer(String extensionName, PropertyMapKeyReplacer replacer) {
+		propertyMapKeyReplacers.put(extensionName, replacer);
+	}
+
+	/**
+	 * Returns by default {*}.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusSearchContext.java]: Try to improve it, please tell me if it's enough.

[the remark this answers is not in the document]
Comment (datho7561) [reply quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusSearchContext.java]: Yes, it look good now.

**Comment by datho7561** [comment]
Sometimes I run into a case where I set up an endpoint that caches a result and confirm the caching is working (i.e. there is no delay on the second request). However, I still don't get auto-completion for the cache properties with the cache name filled. When this is happening, sometimes it gives no cache related properties, and sometimes it gives the default `{*}` property. I'll try to come up with steps to reproduce this issue.

**Comment by angelozerr** [comment]
Ate you using my patch https://github.com/eclipse/lsp4mp/pull/159 ?

**Comment by datho7561** [comment]
Yes, I double checked that I was using the latest lsp4mp, which has your changes in it.

**Comment by datho7561** [comment]
I can't reproduce it reliably, so I think its probably okay to ignore for now.

**Review by datho7561 — APPROVED** [review]

**Comment by angelozerr** [comment]
> I can't reproduce it reliably, so I think its probably okay to ignore for now.

Thanks for the merge. It's annoying that you cannot reproduce the problem. I will try to play with the feature and try to reproduce the problem.

**Pull request merged**